src/services/rfqm_service.ts

```diff
@@ -130,4 +156,97 @@ export class RfqmService {
             allowanceTarget: this._contractAddresses.exchangeProxy,
         };
     }
+
+    /**
+     * Fetch the best firm quote available, including a metatransaction. Returns null if no valid quotes found
+     */
+    public async fetchFirmQuoteAsync(params: FetchFirmQuoteParams): Promise<FetchFirmQuoteResponse | null> {
+        // Extract params
+        const {
+            sellAmount,
+            buyAmount,
+            sellToken: takerToken,
+            buyToken: makerToken,
+            sellTokenDecimals: takerTokenDecimals,
+            buyTokenDecimals: makerTokenDecimals,
+            gasPrice: inputGasPrice,
+            apiKey,
+            takerAddress,
+        } = params;
+
+        // Quote Requestor specific params
+        const isSelling = sellAmount !== undefined;
+        const marketOperation = isSelling ? MarketOperation.Sell : MarketOperation.Buy;
+        const assetFillAmount = isSelling ? sellAmount! : buyAmount!;
+
+        // Prepare gas estimate
+        const gas: BigNumber =
+            inputGasPrice !== undefined
+                ? inputGasPrice
+                : await this._protocolFeeUtils.getGasPriceEstimationOrThrowAsync();
+
+        // Fetch quotes
+        const opts = {
+            ...RFQM_DEFAULT_OPTS,
+            takerAddress,
+            txOrigin: this._registryAddress,
+            apiKey,
+            intentOnFilling: true,
+            isIndicative: false,
+        };
+        const firmQuotes = await this._quoteRequestor.requestRfqmFirmQuotesAsync(
+            makerToken,
+            takerToken,
+            assetFillAmount,
+            marketOperation,
+            undefined,
+            opts,
+        );
+
+        // Get the best quote
+        const bestQuote = getBestQuote(
+            firmQuotes,
+            isSelling,
+            takerToken,
+            makerToken,
+            assetFillAmount,
+            RFQM_MINUMUM_EXPIRY_DURATION_MS,
+        );
+
+        // No quote found
+        if (bestQuote === null) {
+            return null;
+        }
+
+        // Generate the Meta Transaction and its hash
+        const rfqOrder = bestQuote.order as RfqOrder;
+        const metaTransaction = this._blockchainUtils.generateMetaTransaction(
+            rfqOrder,
+            bestQuote.signature,
+            takerAddress,
+            bestQuote.order.takerAmount,
+            CHAIN_ID,
+        );
+        const metaTransactionHash = metaTransaction.getHash();
```

Comment:
Is it worth checking here that the metaTransaction is fillable?

Once the taker signs and submits, they'll expect a fill.

Reply:
Yes, you're right, we should check early. This also opens a small question, which is, if the metaTransaction is not fillable and there are other metaTransactions from MMs, should we try the others as fallbacks? How many fallbacks to consider?

Will see if there's a simple solution that can be coded up